lib/req/steps.ex

```diff
@@ -351,6 +351,7 @@ defmodule Req.Steps do
         Regex.replace(~r/:([a-zA-Z]{1}[\w_]*)/, path, fn match, key ->
           to_string(params[String.to_existing_atom(key)] || match)
         end)
+        |> URI.encode()
```

Comment:
Maybe we just URI.encode the parts we're dynamically replacing instead of the whole thing? If someone ends up relying on this step to fix up their otherwise broken URL it would be really surprising when things get broken when they stop using path params.

In fact, and really sorry being still [at it](https://github.com/wojtekmach/req/pull/164#discussion_r1085001422), but I'm still slightly bothered about the string->atom conversion and it didn't occur to me know that usually the answer is to do the atom->string conversion:

```elixir
params = Map.new(params, fn {key, value} -> {Atom.to_string(key), value} end)

Regex.replace(~r/:([a-zA-Z]{1}[\w_]*)/, path, fn match, key ->
  case Map.fetch(params, key) do
    {:ok, value} -> URI.encode(to_string(value))
    :error -> match
  end
end)
```

wdyt?

Reply:
First part sounds good to me.

Second part, we're trading time complexity for memory complexity given the cost of strings and creating a new map. I put together a benchmark together for both approaches and the proposed method appears to be marginally slower and consumes 12% more memory. This leads me to just question what the benefit is of changing the approach.

```
Operating System: macOS
CPU Information: Apple M1 Pro
Number of Available Cores: 10
Available memory: 32 GB
Elixir 1.14.2
Erlang 24.2

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
reduction time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 56 s

Benchmarking string_to_atom ...
Benchmarking string_to_atom_2 ...
Benchmarking with_map_conversion ...
Benchmarking with_map_conversion_2 ...

Name                            ips        average  deviation         median         99th %
string_to_atom               238.92        4.19 ms     ±1.62%        4.17 ms        4.42 ms
with_map_conversion          232.20        4.31 ms     ±1.76%        4.29 ms        4.57 ms
string_to_atom_2             125.19        7.99 ms     ±1.93%        7.98 ms        8.46 ms
with_map_conversion_2        121.82        8.21 ms     ±1.81%        8.20 ms        8.66 ms

Comparison:
string_to_atom               238.92
with_map_conversion          232.20 - 1.03x slower +0.121 ms
string_to_atom_2             125.19 - 1.91x slower +3.80 ms
with_map_conversion_2        121.82 - 1.96x slower +4.02 ms

Memory usage statistics:

Name                     Memory usage
string_to_atom                1.98 MB
with_map_conversion           2.23 MB - 1.12x memory usage +0.24 MB
string_to_atom_2              3.54 MB - 1.78x memory usage +1.56 MB
with_map_conversion_2         3.94 MB - 1.99x memory usage +1.96 MB

**All measurements for memory usage were the same**
```

```elixir
Mix.install([{:benchee, "~> 1.1"}])

defmodule Test do
  def params, do: %{user_id: "ef400730-94c5-4375-b49d-1e65922dfeb8"}
  def params_2, do: %{user_id: "ef400730-94c5-4375-b49d-1e65922dfeb8", resource_id: "df873829-7be8-4acc-9842-6065b43dc7fe"}

  def with_string_to_atom(path, params) do
    Regex.replace(~r/:([a-zA-Z]{1}[\w_]*)/, path, fn match, key ->
      to_string(params[String.to_existing_atom(key)] || match)
      |> URI.encode()
    end)
  end

  def with_map_conversion(path, params) do
    params = Map.new(params, fn {key, value} -> {Atom.to_string(key), value} end)

    Regex.replace(~r/:([a-zA-Z]{1}[\w_]*)/, path, fn match, key ->
      case Map.fetch(params, key) do
        {:ok, value} -> URI.encode(to_string(value))
        :error -> match
      end
    end)
  end
end

Enum.each(1..100_000, fn n -> Integer.to_string(n) |> String.to_atom() end)

paths = Enum.map(1..1000, fn _ -> "users/:user_id" end)
paths_2 = Enum.map(1..1000, fn _ -> "resources/:resource_id/users/:user_id" end)

Benchee.run(
  %{
    "string_to_atom" => fn -> Enum.map(paths, fn path -> Test.with_string_to_atom(path, Test.params()) end) end,
    "with_map_conversion" => fn -> Enum.map(paths, fn path -> Test.with_map_conversion(path, Test.params()) end) end,
    "string_to_atom_2" => fn -> Enum.map(paths_2, fn path -> Test.with_string_to_atom(path, Test.params_2()) end) end,
    "with_map_conversion_2" => fn -> Enum.map(paths_2, fn path -> Test.with_map_conversion(path, Test.params_2()) end) end
  },
  time: 10,
  memory_time: 2
)
```